# Credit Risk Analysis

### Importing Libraries

In [1]:
import os
import sys
import subprocess
import pandas as pd
import importlib
import warnings
warnings.filterwarnings("ignore", category=ImportWarning)

project_root = os.path.abspath(os.path.join(os.getcwd(), ".."))
sys.path.append(project_root)

In [2]:
from src.train import train_model, clean_feature_names
from src.predict import predict_model
from src.data_load import load_data

### Load and Inspect Data

In [3]:
data_path = "/Users/elbethelzewdie/Downloads/credit-risk-analysis/Credit-risk-analysis/data/processed/final_dataset_new.csv"
df = load_data(data_path)
df.head()

CustomerId  Amount_sum   Amount_avg  Amount_count   Amount_std  \
0  CustomerId_1001      5000.0  2500.000000             2  3818.376618   
1  CustomerId_1002      4225.0   384.090909            11   560.498966   
2  CustomerId_1003      5000.0  1666.666667             3  2753.785274   
3  CustomerId_1004      2000.0  2000.000000             1          NaN   
4  CustomerId_1005      3800.0  1266.666667             3  3257.811126   

   Value_sum    Value_avg  Value_count    Value_std  ProviderId_mode  ...  \
0       5400  2700.000000            2  3535.533906                3  ...   
1       4775   434.090909           11   518.805446                3  ...   
2       7000  2333.333333            3  1892.969449                5  ...   
3       2000  2000.000000            1          NaN                0  ...   
4       6200  2066.666667            3  2571.640203                3  ...   

   ChannelId_mode  transaction_hour_avg  transaction_day_avg  \
0               1              8.000000            16.000000   
1               1             13.454545            14.727273   
2               2             14.666667             1.000000   
3               2             18.000000            22.000000   
4               1              9.666667            28.000000   

   transaction_month_avg  transaction_year_avg  transaction_dayofweek_avg  \
0              11.000000           2018.000000                   4.000000   
1               8.636364           2018.272727                   2.909091   
2               2.000000           2019.000000                   4.000000   
3              12.000000           2018.000000                   5.000000   
4              12.000000           2018.000000                   4.000000   

   is_weekend_avg  is_business_hours_avg  FraudResult  is_high_risk  
0        0.000000               0.000000            0             1  
1        0.181818               0.454545            0             1  
2        0.000000               1.000000            0             1  
3        1.000000               0.000000            0             1  
4        0.000000               0.333333            0             1  

[5 rows x 22 columns]

In [4]:
# Ensure your dataset has ID and target
print("Columns:", df.columns.tolist())

TARGET_COL = "is_high_risk"
ID_COL = "CustomerId"

Columns: ['CustomerId', 'Amount_sum', 'Amount_avg', 'Amount_count', 'Amount_std', 'Value_sum', 'Value_avg', 'Value_count', 'Value_std', 'ProviderId_mode', 'ProductId_mode', 'ProductCategory_mode', 'ChannelId_mode', 'transaction_hour_avg', 'transaction_day_avg', 'transaction_month_avg', 'transaction_year_avg', 'transaction_dayofweek_avg', 'is_weekend_avg', 'is_business_hours_avg', 'FraudResult', 'is_high_risk']


### Checking for Missing Values

In [5]:
print(df.isna().sum())

CustomerId                     0
Amount_sum                     0
Amount_avg                     0
Amount_count                   0
Amount_std                   733
Value_sum                      0
Value_avg                      0
Value_count                    0
Value_std                    733
ProviderId_mode                0
ProductId_mode                 0
ProductCategory_mode           0
ChannelId_mode                 0
transaction_hour_avg           0
transaction_day_avg            0
transaction_month_avg          0
transaction_year_avg           0
transaction_dayofweek_avg      0
is_weekend_avg                 0
is_business_hours_avg          0
FraudResult                    0
is_high_risk                   0
dtype: int64


In [6]:
df['Amount_std'] = df['Amount_std'].fillna(0)
df['Value_std'] = df['Value_std'].fillna(0)


In [7]:
print(df.isna().sum())

CustomerId                   0
Amount_sum                   0
Amount_avg                   0
Amount_count                 0
Amount_std                   0
Value_sum                    0
Value_avg                    0
Value_count                  0
Value_std                    0
ProviderId_mode              0
ProductId_mode               0
ProductCategory_mode         0
ChannelId_mode               0
transaction_hour_avg         0
transaction_day_avg          0
transaction_month_avg        0
transaction_year_avg         0
transaction_dayofweek_avg    0
is_weekend_avg               0
is_business_hours_avg        0
FraudResult                  0
is_high_risk                 0
dtype: int64


### Defining Features and Target Variable

In [8]:
# Define features and target
X = df.drop(columns=[TARGET_COL, ID_COL])
y = df[TARGET_COL]

### Train All Models

In [9]:
print("🚀 Training models...")

metrics = train_model(X, y)

print("\n🎉 Training complete. Evaluation metrics:")

for model_name, metric_vals in metrics.items():
    print(f"\n{model_name}:")
    for k, v in metric_vals.items():
        print(f"  {k}: {v:.4f}")


🚀 Training models...


2026/02/19 12:10:05 INFO alembic.runtime.plugins: setup plugin alembic.autogenerate.schemas
2026/02/19 12:10:05 INFO alembic.runtime.plugins: setup plugin alembic.autogenerate.tables
2026/02/19 12:10:05 INFO alembic.runtime.plugins: setup plugin alembic.autogenerate.types
2026/02/19 12:10:05 INFO alembic.runtime.plugins: setup plugin alembic.autogenerate.constraints
2026/02/19 12:10:05 INFO alembic.runtime.plugins: setup plugin alembic.autogenerate.defaults
2026/02/19 12:10:05 INFO alembic.runtime.plugins: setup plugin alembic.autogenerate.comments
2026/02/19 12:10:05 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/02/19 12:10:05 INFO mlflow.store.db.utils: Updating database tables
2026/02/19 12:10:05 INFO alembic.runtime.migration: Context impl SQLiteImpl.
2026/02/19 12:10:05 INFO alembic.runtime.migration: Will assume non-transactional DDL.
2026/02/19 12:10:05 INFO alembic.runtime.migration: Running upgrade  -> 451aebb31d03, add metric step
2026/02/19 12:1

✅ Logistic Regression model saved to LogisticRegression_best_model.pkl


/Users/elbethelzewdie/Downloads/credit-risk-analysis/Credit-risk-analysis/week12/lib/python3.10/site-packages/mlflow/models/model.py:1209: FutureWarning: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization.The recommended safe alternative is the 'skops' format.
  flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
Successfully registered model 'LogisticRegression'.
Created version '1' of model 'LogisticRegression'.
2026/02/19 12:10:16 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
/Users/elbethelzewdie/Downloads/credit-risk-analysis/Credit-risk-analysis/week12/lib/python3.10/site-packages/mlflow/models/model.py:1209: FutureWarning: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's obje


🎉 Training complete. Evaluation metrics:

LogisticRegression:
  accuracy: 0.9985
  precision: 0.9985
  recall: 1.0000
  f1_score: 0.9993
  roc_auc: 0.9882

DecisionTree:
  accuracy: 0.9971
  precision: 0.9985
  recall: 0.9985
  f1_score: 0.9985
  roc_auc: 0.4993

RandomForest:
  accuracy: 0.9985
  precision: 0.9985
  recall: 1.0000
  f1_score: 0.9993
  roc_auc: 0.9412

XGBoost:
  accuracy: 0.9985
  precision: 0.9985
  recall: 1.0000
  f1_score: 0.9993
  roc_auc: 0.9868


Successfully registered model 'XGBoost'.
Created version '1' of model 'XGBoost'.


- Based on the ROC-AUC scores, Logistic Regression achieved the highest discriminative performance (0.9882), slightly outperforming XGBoost (0.9868) and Random Forest (0.9412). Decision Tree performed poorly in distinguishing classes, with an ROC-AUC near 0.5, indicating it is almost random despite high accuracy due to class imbalance. Overall, Logistic Regression is the best model for this dataset in terms of ROC-AUC.

In [10]:
# Prepare features for prediction (exclude target & ID)
X_pred = df.drop(columns=[TARGET_COL, ID_COL])

# Apply the same preprocessing
X_pred_processed = clean_feature_names(X_pred)

### Best Model for Predictions

In [11]:
best_model_name = max(metrics, key=lambda m: metrics[m]["roc_auc"])
print(f"\n🏆 Best model based on ROC-AUC: {best_model_name}")


🏆 Best model based on ROC-AUC: LogisticRegression


In [12]:
# Predict using the best model
predictions = predict_model(model_name=best_model_name, input_data=X_pred)

# Combine predictions with IDs for reference
result_df = pd.DataFrame({
    ID_COL: df[ID_COL],
    "predicted_target": predictions
})
result_df.head()

# Optionally save predictions
result_df.to_csv("/Users/elbethelzewdie/Downloads/credit-risk-analysis/Credit-risk-analysis/data/processed/predictions.csv", index=False)
print("\n✅ Predictions saved to data/predictions.csv")


✅ Predictions saved to data/predictions.csv
